### PR TITLE
Blog about javascript SDKs

### PR DESCRIPTION
**Pull request by mooreds**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205748704623040
  - https://app.asana.com/0/0/1205750380835987

### CODE REVIEW

File: astro/src/content/blog/javascript-sdks.mdx [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+* JavaScript single pages applications (SPAs)
+* traditional webapps
+* commercial off the shelf applications
+* server side applications
+* APIs
+* mobile apps
+
+Pretty much anything which can use OpenID Connect (OIDC), Security Assertion Markup Language (SAML), or an HTTP based API should work. You can see our [list of quickstarts](/docs/quickstarts/) for integration examples.
+
+Recently, the team built software development kits (SDKs) to make it even easier for SPA developers. The SDKs support the following JavaScript front end frameworks:
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:29]:
```suggestion
Recently, the team built software development kits (SDKs) to make it even easier for SPA developers. The SDKs support the following JavaScript front-end frameworks:
```

File: astro/src/content/blog/javascript-sdks.mdx [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+This is the recommended approach for two reasons:
+
+* more extensive functionality
+* increased security
+
+Let's discuss each of these in turn.
+
+### Functionality
+
+When you use the Authorization Code grant for authentication, as well as redirects for registration and logout, you are leveraging FusionAuth provided functionality. Your application immediately supports a large number of authentication workflows. These include, but are not limited to:
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:119]:
```suggestion
When you use the Authorization Code grant for authentication, as well as redirects for registration and logout, you are leveraging the FusionAuth provided functionality. Your application immediately supports a large number of authentication workflows. These include, but are not limited to:
```

Comment (alex-fusionauth) [reply astro/src/content/blog/javascript-sdks.mdx:119]: Should we add backlinks to these?

File: astro/src/content/blog/javascript-sdks.mdx [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+* social sign-on (Google, Facebook, etc)
+* profile management (with a paid license)
+* passkeys (with a paid license)
+* and more
+
+Now, it's not entirely effortless. You must configure FusionAuth, but all this functionality is present, tested, and maintained over time. For example, instead of having to figure out how to set up email verification, you configure a customizable email and enable it in FusionAuth, and it works. In addition, browser redirects work with any language, framework or browser. This ensures your application's login page is never a barrier to your users.
+
+### Security
+
+When using a redirect, the only application which sees or holes sensitive user credentials is FusionAuth. Your application code never has access to these, which decreases the attack surface area.
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:136]:
```suggestion
When using a redirect, the only application that sees or holes sensitive user credentials is FusionAuth. Your application code never has access to these, which decreases the attack surface area.
```

File: astro/src/content/blog/javascript-sdks.mdx [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+* rate limiting of sensitive security related actions (with an Enterprise license)
+* and more
+
+[FusionAuth takes security seriously](/security); it's our full time job. By redirecting to FusionAuth, you don't have to worry about any other JavaScript in your application getting access to sensitive information or credentials.
+
+### Look and Feel
+
+Additional functionality and security are great, but you typically want your login pages to look like your application. Doesn't a browser redirect make that difficult?
+
+Well, no. FusionAuth allows complete control over the look and feel of the hosted login pages, including adding or removing HTML, CSS, images and JavaScript to the pages. Learn more about this functionality, which we call [themes](/docs/v1/tech/themes/), including an [example using Tailwind CSS](/docs/v1/tech/themes/tailwind).
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:153]:
```suggestion
Well, no. FusionAuth allows complete control over the look and feel of the hosted login pages, including adding or removing HTML, CSS, images, and JavaScript to the pages. Learn more about this functionality, which we call [themes](/docs/v1/tech/themes/), including an [example using Tailwind CSS](/docs/v1/tech/themes/tailwind).
```

File: astro/src/content/blog/javascript-sdks.mdx [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+If your application architecture, use case, or design team rules out redirects, you can use the [Login API](/docs/v1/tech/apis/login) to directly pass user credentials to FusionAuth. 
+
+You'll miss out on the functionality and security benefits outlined above, however.
+
+At the current time there's no SDK support for this approach, but FusionAuth users can and have built out their own user interface using the Login API.
+
+## In Conclusion
+
+Using the Angular, React or Vue SDKs can accelerate your app's integration with FusionAuth.
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:175]:
```suggestion
Using the Angular, React, or Vue SDKs can accelerate your app's integration with FusionAuth.
```

File: astro/src/content/blog/javascript-sdks.mdx [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+By doing so, you can focus on the features you users actually care about while leveraging a world-class implementation of undifferentiated login workflows and security of an isolated authentication server.
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:177]:
```suggestion
By doing so, you can focus on the features your users actually care about while leveraging a world-class implementation of undifferentiated login workflows and the security of an isolated authentication server.
```

File: astro/src/content/blog/javascript-sdks.mdx [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+## Client Libraries vs SDKs
+
+At FusionAuth, a client library is different from an SDK, even though many people in the industry use the terms interchangeably.
+
+### Client Libraries
+
+A client library is a wrapper over [FusionAuth APIs](/docs/v1/tech/apis/). Client libraries support all public FusionAuth API endpoints, and offer a convenient way to use that API in many software languages. At time of writing, there are eight supported languages, including TypeScript. There's also an OpenAPI definition, useful for viewing API changes over time or generating a client library for an unsupported language like C++ or Rust.
+
+If you want to compare a client library to a child's toy, it is like tinker toys. Engineers build a script or program using a client library to extend or configure FusionAuth. Some examples of tasks well suited to a client library:
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:45]: I don't get the analogy here. Maybe add that like tinker toys you have the building blocks but not the complete toy. The millennial in me also wouldn't reference a tinker toy...but that is for another day.

File: astro/src/content/blog/javascript-sdks.mdx [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+### Look and Feel
+
+Additional functionality and security are great, but you typically want your login pages to look like your application. Doesn't a browser redirect make that difficult?
+
+Well, no. FusionAuth allows complete control over the look and feel of the hosted login pages, including adding or removing HTML, CSS, images and JavaScript to the pages. Learn more about this functionality, which we call [themes](/docs/v1/tech/themes/), including an [example using Tailwind CSS](/docs/v1/tech/themes/tailwind).
+
+One other common objection to redirection is the ability to send a successfully authenticated user right back to where they started. It's common to allow users to log in from multiple states in your SPA, after all. Luckily, you can use the `state` parameter to deep link from any page in your SPA to FusionAuth and then send the user right back to that page after a successful login. Here's a video displaying that:
+
+<YouTube id="-vx5rdy-mvY" />
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:157]: Let's have Brand/Design get a cover on this one too! I don't think we should show any videos without those.

File: astro/src/content/blog/javascript-sdks.mdx [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+The takeaway from this code is that you can set the `state` parameter when redirecting the user to the login page, then extract a value from it and redirect back to the correct location.
+
+### Redirects Don't Work, Help!
+
+If your application architecture, use case, or design team rules out redirects, you can use the [Login API](/docs/v1/tech/apis/login) to directly pass user credentials to FusionAuth. 
+
+You'll miss out on the functionality and security benefits outlined above, however.
+
+At the current time there's no SDK support for this approach, but FusionAuth users can and have built out their own user interface using the Login API.
```
[the remark this answers is not in the document]
Comment (alex-fusionauth) [reply astro/src/content/blog/javascript-sdks.mdx:171]: Do we have any examples we could link to on this?

**Review by alex-fusionauth — CHANGES_REQUESTED** [review]
Technical looks good (love that we have that remotecode component).

Couple spelling things that I would fix up to you.

File: astro/src/content/blog/javascript-sdks.mdx (hunk 2 above)
Comment (mooreds) [reply astro/src/content/blog/javascript-sdks.mdx:119]: Probably. Good call.

File: astro/src/content/blog/javascript-sdks.mdx (hunk 8 above)
Comment (mooreds) [reply astro/src/content/blog/javascript-sdks.mdx:157]: Great, will add a task.

File: astro/src/content/blog/javascript-sdks.mdx (hunk 9 above)
[the remark this answers is not in the document]
Comment (mooreds) [reply astro/src/content/blog/javascript-sdks.mdx:171]: No, it's not really recommended. I actually wasn't sure I should include it, but it is a supported use case.

File: astro/src/content/blog/javascript-sdks.mdx [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+* more extensive functionality
+* increased security
+
+Let's discuss each of these in turn.
+
+### Functionality
+
+When you use the Authorization Code grant for authentication, as well as redirects for registration and logout, you are leveraging the FusionAuth provided functionality. Your application immediately supports a large number of authentication workflows. These include, but are not limited to:
+
```
Comment (mooreds) [inline astro/src/content/blog/javascript-sdks.mdx:120]: Add links here.

File: astro/src/content/blog/javascript-sdks.mdx (hunk 7 above)
Comment (mooreds) [reply astro/src/content/blog/javascript-sdks.mdx:45]: I could use legos as the example too?

File: astro/src/content/blog/javascript-sdks.mdx [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+* forgot password
+* email verification
+* [MFA](/docs/v1/tech/guides/multi-factor-authentication)
+* [magic link authentication](/docs/v1/tech/passwordless/magic-links)
+* password expiration
+* [enterprise single sign-on (SAML, OIDC)](/docs/v1/tech/identity-providers/#enterprise-identity-providers)
+* [social sign-on (Google, Facebook, etc)](http://localhost:4000/docs/v1/tech/identity-providers/#social-identity-providers)
```
Comment (bhalsey) [inline astro/src/content/blog/javascript-sdks.mdx:127]:
```suggestion
* [social sign-on (Google, Facebook, etc)](/docs/v1/tech/identity-providers/#social-identity-providers)
```

File: astro/src/content/blog/javascript-sdks.mdx [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+import { YouTube } from '@astro-community/astro-embed-youtube';
+import RemoteCode from "../../components/RemoteCode.astro";
+
+FusionAuth is, at its core, an architectural component of a larger application. I like to say "no one ever builds a full application with just FusionAuth". Sure, your application needs authentication and authorization, but there are other critical components. For example, the features of your application which make users *want* to register and log in.
+
+{/* more */}
+
+But here at FusionAuth, we want simple, secure authentication to be easy to add to any application. FusionAuth can run anywhere, but what use is that if it isn't easy to integrate? Applications which FusionAuth works with include:
+
+* JavaScript single pages applications (SPAs)
```
Comment (bhalsey) [inline astro/src/content/blog/javascript-sdks.mdx:20]:
```suggestion
* JavaScript single page applications (SPAs)
```

File: astro/src/content/blog/javascript-sdks.mdx [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+This post will cover the nitty gritty of using these SDKs, but first, let's discuss the difference between an SDK and a [client library](/docs/v1/tech/client-libraries/).
+
+## Client Libraries vs SDKs
+
+At FusionAuth, a client library is different from an SDK, even though many people in the industry use the terms interchangeably.
+
+### Client Libraries
+
+A client library is a wrapper over [FusionAuth APIs](/docs/v1/tech/apis/). Client libraries support all public FusionAuth API endpoints and offer a convenient way to use that API in many software languages. At the time of writing, there are eight supported languages, including TypeScript. There's also an OpenAPI definition, useful for viewing API changes over time or generating a client library for an unsupported language like C++ or Rust.
```
Comment (bhalsey) [inline astro/src/content/blog/javascript-sdks.mdx:43]:
```suggestion
A client library is a wrapper over [FusionAuth APIs](/docs/v1/tech/apis/). Client libraries support all public FusionAuth API endpoints and offer a convenient way to use that API in many software languages. At the time of writing, there are eight supported languages, including TypeScript. There's also an OpenAPI definition, useful for viewing API changes over time or making HTTP calls from an unsupported language like C++ or Rust.
```

**Review by bhalsey — COMMENTED** [review]
I'd consider flipping the presentation order:
* First the client side SDK
* then the *optional* server side component of the SDK
* then the client library for customers needing even more customizability

File: astro/src/content/blog/javascript-sdks.mdx [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+authors: Dan Moore
+excerpt_separator: "{/* more */}"
+image: /img/blogs/javascript-sdks/our-new-vue-react-and-angular-sdks.png
+categories: Company
+tags: javascript, sdks, angular, vue, react, client libraries
+---
+import { YouTube } from '@astro-community/astro-embed-youtube';
+import RemoteCode from "../../components/RemoteCode.astro";
+
+FusionAuth is, at its core, an architectural component of a larger application. I like to say "no one ever builds a full application with just FusionAuth". Sure, your application needs authentication and authorization, but there are other critical components. For example, the features of your application which make users *want* to register and log in.
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:14]:
```suggestion
FusionAuth is, at its core, an architectural component of a larger application. I like to say "no one ever builds a full application with just FusionAuth". Sure, your application needs authentication and authorization, but there are other critical components. For example, the features of your application that make users *want* to register and log in.
```

File: astro/src/content/blog/javascript-sdks.mdx [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+If you want to compare a client library to a child's toy, it is like legos. Engineers build a script or program using a client library to extend or configure FusionAuth. Some examples of tasks well suited to a client library:
+
+* [rotating client secrets or API keys](/docs/v1/tech/tutorials/key-rotation) to improve your security posture
+* pulling user data every night for data warehouse ingestion
+* creating a new FusionAuth tenant every time a client signs up for your app
+* offering a unique, custom login or registration experience not supported by the [FusionAuth hosted login pages](/docs/v1/tech/core-concepts/integration-points#hosted-login-pages)
+
+Requests made with client libraries usually require an API key as well as a URL for the FusionAuth instance which the requests will be applied to.
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:66]:
```suggestion
Requests made with client libraries usually require an API key as well as a URL for the FusionAuth instance to which the requests will be applied.
```

File: astro/src/content/blog/javascript-sdks.mdx [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+Additional functionality and security are great, but you typically want your login pages to look like your application. Doesn't a browser redirect make that difficult?
+
+Well, no. FusionAuth allows complete control over the look and feel of the hosted login pages, including adding or removing HTML, CSS, images, and JavaScript to the pages. Learn more about this functionality, which we call [themes](/docs/v1/tech/themes/), including an [example using Tailwind CSS](/docs/v1/tech/themes/tailwind).
+
+One other common objection to redirection is the ability to send a successfully authenticated user right back to where they started. It's common to allow users to log in from multiple states in your SPA, after all. Luckily, you can use the `state` parameter to deep link from any page in your SPA to FusionAuth and then send the user right back to that page after a successful login. Here's a video displaying that:
+
+<YouTube id="-vx5rdy-mvY" />
+
+Here's code from an [example application](https://github.com/FusionAuth/fusionauth-example-node-deeplink/) which extracts a URL from a `state` parameter and returns the user to that URL after a successful exchange at the token endpoint.
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:150]:
```suggestion
Here's code from an [example application](https://github.com/FusionAuth/fusionauth-example-node-deeplink/) that extracts a URL from a `state` parameter and returns the user to that URL after a successful exchange at the token endpoint.
```

File: astro/src/content/blog/javascript-sdks.mdx [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+For SPAs which must request APIs that don't live on the same domain the SPA is served from, cookies won't work. In this situation, the backend for frontend pattern is recommended. **You should not put bearer tokens into local storage or session storage.** Here's a video about the insecurity of storing tokens anyplace accessible to JavaScript running in the browser.
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:160]:
```suggestion
For SPAs that must request APIs that don't live on the same domain the SPA is served from, cookies won't work. In this situation, the backend for front-end pattern is recommended. **You should not put bearer tokens into local storage or session storage.** Here's a video about the insecurity of storing tokens anyplace accessible to JavaScript running in the browser.
```

Comment (alex-fusionauth) [reply astro/src/content/blog/javascript-sdks.mdx:160]: Maybe this should even be "Back-end for Front-end" not sure on that one.

File: astro/src/content/blog/javascript-sdks.mdx [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+* [rotating client secrets or API keys](/docs/v1/tech/tutorials/key-rotation) to improve your security posture
+* pulling user data every night for data warehouse ingestion
+* creating a new FusionAuth tenant every time a client signs up for your app
+* offering a unique, custom login or registration experience not supported by the [FusionAuth hosted login pages](/docs/v1/tech/core-concepts/integration-points#hosted-login-pages)
+
+Requests made with client libraries usually require an API key as well as a URL for the FusionAuth instance which the requests will be applied to.
+
+### Better Together?
+
+You can use both an SDK and a client library if your use case requires the front-end SDK-based solutions for easy integration and the management access provided by a client library. They are not mutually exclusive.
```
Comment (alex-fusionauth) [inline astro/src/content/blog/javascript-sdks.mdx:70]:
```suggestion
You can use both an SDK and a client library if your use case requires front-end SDK-based solutions for easy integration and the management access provided by a client library. They are not mutually exclusive.
```

**Review by alex-fusionauth — APPROVED** [review]

**Pull request merged**